Source/WebCore/html/ImageDataArray.cpp

```diff
@@ -29,6 +29,185 @@
+    if (auto* array = dynamicDowncast<JSC::Uint8ClampedArray>(bufferView))
+        return std::forward<F>(f)(*array);
+    if (auto* array = dynamicDowncast<JSC::Float16Array>(bufferView))
+        return std::forward<F>(f)(*array);
+    RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+}
+
+ImageDataArray::ImageDataArray(Ref<JSC::ArrayBufferView>&& arrayBufferView)
+    : m_arrayBufferView(WTFMove(arrayBufferView))
+{
+    ASSERT(isSupported(m_arrayBufferView.get()));
+}
+
+ImageDataArray::ImageDataArray(Ref<JSC::Uint8ClampedArray>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
+
+ImageDataArray::ImageDataArray(Ref<JSC::Float16Array>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
+
+std::optional<ImageDataArray> ImageDataArray::tryCreate(size_t length, ImageDataStorageFormat storageFormat, bool zeroFill)
+{
+    std::optional<ImageDataArray> array;
+    switch (storageFormat) {
+    case ImageDataStorageFormat::Uint8:
+        if (RefPtr bufferView = JSC::Uint8ClampedArray::tryCreateUninitialized(length)) {
+            if (zeroFill)
+                bufferView->zeroFill();
+            array.emplace(bufferView.releaseNonNull());
+        }
+        break;
+    case ImageDataStorageFormat::Float16:
+        if (RefPtr bufferView = JSC::Float16Array::tryCreateUninitialized(length)) {
+            if (zeroFill)
+                bufferView->zeroFill();
+            array.emplace(bufferView.releaseNonNull());
+        }
+    }
+    return array;
+}
+
+std::optional<ImageDataArray> ImageDataArray::tryCreateZeroFilled(size_t length, ImageDataStorageFormat storageFormat)
+{
+    return tryCreate(length, storageFormat, true);
+}
+
+std::optional<ImageDataArray> ImageDataArray::tryCreateFromBufferOrZero(size_t length, ImageDataStorageFormat storageFormat, const uint8_t* buffer, size_t bytes)
+{
+    std::optional<ImageDataArray> array = tryCreate(length, storageFormat, false);
+    if (array) {
+        visitArrayBufferView(
+            array->arrayBufferView(),
+            [&](auto& array) {
+                memcpy(array.data(), buffer, bytes);
+            });
+    }
+    return array;
+}
+
+constexpr bool ImageDataArray::isSupported(JSC::TypedArrayType typedArrayType)
+{
+    switch (typedArrayType) {
+    case JSC::TypeUint8Clamped:
+    case JSC::TypeFloat16:
+        return true;
+    default:
+        return false;
+    }
+}
+
+constexpr bool ImageDataArray::isSupported(const JSC::ArrayBufferView& arrayBufferView)
+{
+    return isSupported(arrayBufferView.getType());
+}
+
+ImageDataStorageFormat ImageDataArray::storageFormat() const
+{
+    switch (m_arrayBufferView->getType()) {
+    case JSC::TypeUint8Clamped: return ImageDataStorageFormat::Uint8;
+    case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+    default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+}
```

Comment:
Should the conversion from TypedArrayType to ImageDataStorageFormat.h be in ImageDataStorageFormat.h?